

# Tune Practice

## *Data Science Unit 4 Sprint 2 Assignment 3*

# Gridsearch Hyperparameters

In the guided project, you learned how to use sklearn's GridsearchCV and keras-tuner library to tune the hyperparamters of a neural network model. For your module project you'll continue using these two libraries however we are going to make things a little more interesting for you. 

Continue to use TensorFlow Keras & a sample of the [Quickdraw dataset](https://github.com/googlecreativelab/quickdraw-dataset) to build a sketch classification model. The dataset has been sampled to only 10 classes and 10000 observations per class. 



**Don't forgot to switch to GPU on Colab!**

In [32]:
# native python libraries imports 
import math
from time import time

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


# sklearn imports 
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV

# keras imports 
import tensorflow.keras
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense
from kerastuner.tuners import RandomSearch, BayesianOptimization, Sklearn
from kerastuner.engine.hyperparameters import HyperParameters
from tensorflow.keras.activations import relu, sigmoid
from tensorflow.keras.optimizers import Adam, SGD
from tensorflow.keras.utils import get_file

# required for compatibility between sklearn and keras
from tensorflow.keras.wrappers.scikit_learn import KerasClassifier

In [33]:
def load_quickdraw10():
    """
    Fill out this doc string, and comment the code, for practice in writing the kind of code that will get you hired. 
    """
    
    URL_ = "https://github.com/LambdaSchool/DS-Unit-4-Sprint-2-Neural-Networks/blob/main/quickdraw10.npz?raw=true"
    
    path_to_zip = get_file('./quickdraw10.npz', origin=URL_, extract=False)

    data = np.load(path_to_zip)
    
    # normalize your image data
    max_pixel_value = 255
    X = data['arr_0']/max_pixel_value
    Y = data['arr_1']
        
    return train_test_split(X, Y, shuffle=True)

In [34]:
X_train, X_test, y_train, y_test = load_quickdraw10()

In [35]:
X_train.shape

(75000, 784)

In [36]:
y_train.shape

(75000,)

_____

# Experiment 1

## Tune Hyperperameters using Enhanced GridsearchCV 

We are going to use GridsearchCV again to tune a deep learning model however we are going to add some additional functionality to our gridsearch. Specifically, we are going to automate away the generation of how many nodes to use in a layer and how many layers to use in a model! 

By the way, yes, there is a function within a function. Try to not let that bother you. An alternative to this would be to create a class. If you're up for the challenge give it a shot. However, consider this a stretch goal that you come back to after you finish going through this assignment. 


### Objective 

The objective of this experiment is to show you how to automate the generation of layers and layer nodes for the purposes of gridsearch. Up until now, we've been manually selecting the number of layers and layer nodes.

In [37]:
# Function to create model, required for KerasClassifier
def create_model(n_layers,  first_layer_nodes, last_layer_nodes, act_funct ="relu", negative_node_incrementation=True):
    """"
    Returns a complied keras model 
    
    Parameters
    ----------
    n_layers: int 
        number of hidden layers in model 
        To be clear, this excludes the input and output layer.
        
    first_layer_nodes: int
        Number of nodes in the first hidden layer 

    last_layer_nodes: int
        Number of nodes in the last hidden layer (this is the layer just prior to the output layer)
        
     act_funct: string 
         Name of activation function to use in hidden layers (this excludes the output layler)
        
    Returns
    -------
    model: keras object 
    """
    
    def gen_layer_nodes(n_layers, first_layer_nodes, last_layer_nodes, negative_node_incrementation=True):
        """
        Generates and returns the number of nodes in each hidden layer. 
        To be clear, this excludes the input and output layer. 

        Note
        ----
        Number of nodes in each layer is linearly incremented. 
        For example, gen_layer_nodes(5, 500, 100) will generate [500, 400, 300, 200, 100]

        Parameters
        ----------
        n_layers: int
            Number of hidden layers
            This values should be 2 or greater 

        first_layer_nodes: int

        last_layer_nodes: int

        Returns
        -------
        layers: list of ints
            Contains number of nodes for each layer 
        """

        # throws an error if n_layers is less than 2 
        assert n_layers >= 2, "n_layers must be 2 or greater"

        layers = []

        # PROTIP: IF YOU WANT THE NODE INCREMENTATION TO BE SPACED DIFFERENTLY
        # THEN YOU'LL NEED TO CHANGE THE WAY THAT IT'S CALCULATED - HAVE FUN!
        # when set to True number of nodes are decreased for subsequent layers 
        if negative_node_incrementation:
            # subtract this amount from previous layer's nodes in order to increment towards smaller numbers 
            nodes_increment = (last_layer_nodes - first_layer_nodes)/ (n_layers-1)
            
        # when set to False number of nodes are increased for subsequent layers
        else:
            # add this amount from previous layer's nodes in order to increment towards larger numbers 
            nodes_increment = (first_layer_nodes - last_layer_nodes)/ (n_layers-1)

        nodes = first_layer_nodes

        for i in range(1, n_layers+1):

            layers.append(math.ceil(nodes))

            # increment nodes for next layer 
            nodes = nodes + nodes_increment

        return layers
    
    # create model
    model = Sequential()
    
    n_nodes = gen_layer_nodes(n_layers, first_layer_nodes, last_layer_nodes, negative_node_incrementation)
    
    for i in range(1, n_layers):
        if i==1:
            model.add(Dense(first_layer_nodes, input_dim=X_train.shape[1], activation=act_funct))
        else:
            model.add(Dense(n_nodes[i-1], activation=act_funct))
            
            
    # output layer 
    model.add(Dense(10, # 10 unit/neurons in output layer because we have 10 possible labels to predict  
                    activation='softmax')) # use softmax for a label set greater than 2            
    
    # Compile model
    model.compile(loss='sparse_categorical_crossentropy', 
                  optimizer='adam', # adam is a good default optimizer 
                  metrics=['accuracy'])
    
    # do not include model.fit() inside the create_model function
    # KerasClassifier is expecting a complied model 
    return model


## Explore create_model

Let's build a few different models in order to understand how the above code works in practice. 

### Build model 

Use `create_model` to build a model. 

- Set `n_layers = 10` 
- Set `first_layer_nodes = 500`
- Set `last_layer_nodes = 100`
- Set `act_funct = "relu"`
- Make sure that `negative_node_incrementation = True`

In [38]:
# use create_model to create a model 

# YOUR CODE HERE
model = create_model(n_layers=10, first_layer_nodes=500, last_layer_nodes=100, act_funct ="relu", negative_node_incrementation=True)

In [39]:
# run model.summary() and make sure that you understand the model architecture that you just built 
# Notice in the model summary how the number of nodes have been linearly incremented in decreasing values. 
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_2 (Dense)              (None, 500)               392500    
_________________________________________________________________
dense_3 (Dense)              (None, 456)               228456    
_________________________________________________________________
dense_4 (Dense)              (None, 412)               188284    
_________________________________________________________________
dense_5 (Dense)              (None, 367)               151571    
_________________________________________________________________
dense_6 (Dense)              (None, 323)               118864    
_________________________________________________________________
dense_7 (Dense)              (None, 278)               90072     
_________________________________________________________________
dense_8 (Dense)              (None, 234)              

### Build model 

Use `create_model` to build a model. 

- Set `n_layers = 10` 
- Set `first_layer_nodes = 500`
- Set `last_layer_nodes = 100`
- Set `act_funct = "relu"`
- Make sure that `negative_node_incrementation = False`

In [40]:
# use create_model to create a model 

# YOUR CODE HERE
model = create_model(n_layers=10, first_layer_nodes=500, last_layer_nodes=100, act_funct ="relu", negative_node_incrementation=False)

In [41]:
# run model.summary() and make sure that you understand the model architecture that you just built 
# Notice in the model summary how the number of nodes have been linearly incremented in increasing values.
# The output layer must have 10 nodes because there are 10 labels to predict 
model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_12 (Dense)             (None, 500)               392500    
_________________________________________________________________
dense_13 (Dense)             (None, 545)               273045    
_________________________________________________________________
dense_14 (Dense)             (None, 589)               321594    
_________________________________________________________________
dense_15 (Dense)             (None, 634)               374060    
_________________________________________________________________
dense_16 (Dense)             (None, 678)               430530    
_________________________________________________________________
dense_17 (Dense)             (None, 723)               490917    
_________________________________________________________________
dense_18 (Dense)             (None, 767)              

In [42]:
# feel free to play around with parameters to gain additional insight as to how the create_model function works 



Ok, now that we've played around a bit with  `create_model` in order to understand how it works, let's build a much simpler model that we'll be running gridsearches. 

### Build model 

Use `create_model` to build a model. 

- Set `n_layers = 2` 
- Set `first_layer_nodes = 500`
- Set `last_layer_nodes = 100`
- Set `act_funct = "relu"`
- Make sure that `negative_node_incrementation = True`

In [43]:
# use create_model to create a model 

model = create_model(n_layers=2, first_layer_nodes=500, last_layer_nodes=100, act_funct ="relu", negative_node_incrementation=True)

In [44]:
# run model.summary() and make sure that you understand the model architecture that you just built 
model.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_22 (Dense)             (None, 500)               392500    
_________________________________________________________________
dense_23 (Dense)             (None, 10)                5010      
Total params: 397,510
Trainable params: 397,510
Non-trainable params: 0
_________________________________________________________________


In [45]:
# define the grid search parameters
param_grid = {'n_layers': [2, 3],
              'epochs': [3], 
              "first_layer_nodes": [500, 300],
              "last_layer_nodes": [100, 50]
             }

In [46]:
model = KerasClassifier(create_model)

In [47]:
# Create Grid Search
grid = GridSearchCV(estimator=model, 
                    param_grid=param_grid, 
                    n_jobs=-2, 
                    verbose=1, 
                    cv=3)

grid_result = grid.fit(X_train, y_train)

# Report Results
print(f"Best: {grid_result.best_score_} using {grid_result.best_params_}")

means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']

for mean, stdev, param in zip(means, stds, params):
    print(f"Means: {mean}, Stdev: {stdev} with: {param}")

Fitting 3 folds for each of 8 candidates, totalling 24 fits


[Parallel(n_jobs=-2)]: Using backend LokyBackend with 7 concurrent workers.
2021-07-28 14:32:28.272794: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-07-28 14:32:28.273846: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-07-28 14:32:28.679860: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations: 

Epoch 1/3
1563/1563 [==============================] - 13s 8ms/step - loss: 0.6620 - accuracy: 0.8028
Epoch 2/3
1563/1563 [==============================] - 13s 8ms/step - loss: 0.6612 - accuracy: 0.8011
Epoch 2/3
1563/1563 [==============================] - 11s 7ms/step - loss: 0.4496 - accuracy: 0.8662
Epoch 3/3
782/782 [==============================] - 3s 3ms/step - loss: 0.4836 - accuracy: 0.8620
Epoch 1/3
1563/1563 [==============================] - 24s 9ms/step - loss: 0.6887 - accuracy: 0.7947
Epoch 2/3
1563/1563 [==============================] - 14s 9ms/step - loss: 0.4516 - accuracy: 0.8649
Epoch 3/3
1563/1563 [==============================] - 21s 8ms/step - loss: 0.6943 - accuracy: 0.7925
Epoch 2/3
1563/1563 [==============================] - 11s 7ms/step - loss: 0.4719 - accuracy: 0.8603
Epoch 3/3
1563/1563 [==============================] - 11s 7ms/step - loss: 0.4760 - accuracy: 0.8593
Epoch 3/3
782/782 [==============================] - 3s 3ms/step - loss: 0.4725 - acc

782/782 [==============================] - 2s 2ms/step - loss: 0.4659 - accuracy: 0.8648
Epoch 1/3
1563/1563 [==============================] - 8s 4ms/step - loss: 0.6569 - accuracy: 0.7985
Epoch 2/3
1563/1563 [==============================] - 8s 5ms/step - loss: 0.4699 - accuracy: 0.8594
Epoch 3/3
1563/1563 [==============================] - 9s 6ms/step - loss: 0.4491 - accuracy: 0.8621
Epoch 3/3
1563/1563 [==============================] - 9s 6ms/step - loss: 0.4477 - accuracy: 0.8633
Epoch 3/3
1563/1563 [==============================] - 10s 5ms/step - loss: 0.6882 - accuracy: 0.7938
Epoch 2/3
1563/1563 [==============================] - 9s 6ms/step - loss: 0.4485 - accuracy: 0.8635
Epoch 3/3
1563/1563 [==============================] - 8s 5ms/step - loss: 0.4725 - accuracy: 0.8609
Epoch 3/3
1563/1563 [==============================] - 8s 5ms/step - loss: 0.4648 - accuracy: 0.8621
Epoch 3/3
1563/1563 [==============================] - 6s 3ms/step - loss: 0.6526 - accuracy: 0.8002
E

[Parallel(n_jobs=-2)]: Done  24 out of  24 | elapsed:  2.8min finished
2021-07-28 14:35:05.714772: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:176] None of the MLIR Optimization Passes are enabled (registered 2)


Epoch 1/3
2344/2344 [==============================] - 4s 1ms/step - loss: 0.6122 - accuracy: 0.8185
Epoch 2/3
2344/2344 [==============================] - 3s 1ms/step - loss: 0.4228 - accuracy: 0.8739
Epoch 3/3
2344/2344 [==============================] - 3s 1ms/step - loss: 0.3453 - accuracy: 0.8968
Best: 0.865559995174408 using {'epochs': 3, 'first_layer_nodes': 500, 'last_layer_nodes': 100, 'n_layers': 2}
Means: 0.865559995174408, Stdev: 0.0016482652977161257 with: {'epochs': 3, 'first_layer_nodes': 500, 'last_layer_nodes': 100, 'n_layers': 2}
Means: 0.8645600080490112, Stdev: 0.0022789651976011295 with: {'epochs': 3, 'first_layer_nodes': 500, 'last_layer_nodes': 100, 'n_layers': 3}
Means: 0.863426665465037, Stdev: 0.0037553833491735 with: {'epochs': 3, 'first_layer_nodes': 500, 'last_layer_nodes': 50, 'n_layers': 2}
Means: 0.8652399977048238, Stdev: 0.0006424844893858848 with: {'epochs': 3, 'first_layer_nodes': 500, 'last_layer_nodes': 50, 'n_layers': 3}
Means: 0.8643200000127157,

In [48]:
best_model = grid_result.best_estimator_

In [49]:
best_model.get_params()

{'epochs': 3,
 'first_layer_nodes': 500,
 'last_layer_nodes': 100,
 'n_layers': 2,
 'build_fn': <function __main__.create_model(n_layers, first_layer_nodes, last_layer_nodes, act_funct='relu', negative_node_incrementation=True)>}

-----

# Experiment 2

## Benchmark different Optimization Algorithms 

In this section, we are going to use the same model and dataset in order to benchmark 3 different gridsearch approaches: 

- Random Search
- Bayesian Optimization. 
- Brute Force Gridsearch

Our goal in this experiment is two-fold. We want to see which appraoch 

- Scores the highest accuracy
- Has the shortest run time 

We want to see how these 3 gridsearch approaches handle these trade-offs and to give you a sense of those trades offs.

### Trade-offs

`Brute Force Gridsearch` will train a model on every single unique hyperparameter combination, this guarantees that you'll get the highest possible accuracy from your parameter set but your gridsearch might have a very long run-time. 

`Random Search` will randomly sample from your parameter set which, depending on how many samples, the run-time might be significantly cut down but you might or might not sample the parameters that correspond to the heightest possible accuracies. 

`Bayesian Optimization` has a bit of intelligence built into it's search algorithm but you do need to manually select some parameters which greatly influence the model learning outcomes. 

-------
### Build our model

In [50]:
# because gridsearching can take a lot of time and we are bench marking 3 different approaches
# let's build a simple model to minimize run time 

def build_model(hp):
    
    """
    Returns a complied keras model ready for keras-tuner gridsearch algorithms 
    """
    
    model = Sequential()
    
    # hidden layer
    model.add(Dense(units=hp.get('units'),activation=hp.get("activation")))
    
    # output layer
    model.add(Dense(10, activation='softmax'))
    
    model.compile(
        optimizer= tensorflow.keras.optimizers.Adam(hp.get('learning_rate')),
        loss='sparse_categorical_crossentropy',
        metrics=['accuracy'])
    
    return model
  

In [51]:
# build out our hyperparameter dictionary 
hp = HyperParameters()
hp.Int('units', min_value=32, max_value=512, step=32)
hp.Choice('learning_rate',values=[1e-1, 1e-2, 1e-3])
hp.Choice('activation',values=["relu", "sigmoid"])

'relu'

------
# Run the Gridsearch Algorithms 

### Random Search

Be sure to check out the [**docs for Keras-Tuner**](https://keras-team.github.io/keras-tuner/documentation/tuners/). Here you can read about the input parameters for the `RandomSearch` tuner.

In [52]:
# how many unique hyperparameter combinations do we have? 
# HINT: take the product of the number of possible values for each hyperparameter 
# save your answer to n_unique_hparam_combos

# YOUR CODE HERE
n_unique_hparam_combos = (512-32)/32 * 3 * 2 # num units, num lr, num activations
n_unique_hparam_combos

90.0

In [53]:
# how many of these do we want to randomly sample?
# let's pick 25% of n_unique_hparam_combos param combos to sample
# save this number to n_param_combos_to_sample

# YOUR CODE HERE
n_param_combos_to_sample = n_unique_hparam_combos*0.25

In [54]:
random_tuner = RandomSearch(
            build_model,
            objective='val_accuracy',
            max_trials=n_param_combos_to_sample, # number of times to sample the parameter set and build a model 
            seed=1234,
            hyperparameters=hp, # pass in our hyperparameter dictionary
            directory='./keras-tuner-trial',
            project_name='random_search')

In [55]:
# take note of Total elapsed time in print out
random_tuner.search(X_train, y_train,
                    epochs=3,
                    validation_data=(X_test, y_test))

Trial 23 Complete [00h 00m 11s]
val_accuracy: 0.8588399887084961

Best val_accuracy So Far: 0.871720016002655
Total elapsed time: 00h 03m 54s
INFO:tensorflow:Oracle triggered exit


In [56]:
# identify the best score and hyperparamter (should be at the top since scores are ranked)
random_tuner.results_summary()

Results summary
Results in ./keras-tuner-trial/random_search
Showing 10 best trials
Objective(name='val_accuracy', direction='max')
Trial summary
Hyperparameters:
units: 352
learning_rate: 0.001
activation: relu
Score: 0.871720016002655
Trial summary
Hyperparameters:
units: 384
learning_rate: 0.001
activation: relu
Score: 0.8683599829673767
Trial summary
Hyperparameters:
units: 480
learning_rate: 0.001
activation: sigmoid
Score: 0.8635200262069702
Trial summary
Hyperparameters:
units: 448
learning_rate: 0.001
activation: sigmoid
Score: 0.8617600202560425
Trial summary
Hyperparameters:
units: 416
learning_rate: 0.001
activation: sigmoid
Score: 0.8614000082015991
Trial summary
Hyperparameters:
units: 288
learning_rate: 0.001
activation: sigmoid
Score: 0.8588399887084961
Trial summary
Hyperparameters:
units: 224
learning_rate: 0.001
activation: sigmoid
Score: 0.8573200106620789
Trial summary
Hyperparameters:
units: 192
learning_rate: 0.01
activation: sigmoid
Score: 0.8416799902915955
Tria

 ### Results
 
Identify and write the the best performing hyperparamter combination and model score. 
Note that because this is Random Search, multiple runs might have slighly different outcomes. 
 
 

YOUR ANSWER HERE

The best hyperparameters were: 
units: 352
learning_rate: 0.001
activation: relu
Score: 0.871720016002655

------
### Bayesian Optimization

![](https://upload.wikimedia.org/wikipedia/commons/0/02/GpParBayesAnimationSmall.gif)

Be sure to check out the [**docs for Keras-Tuner**](https://keras-team.github.io/keras-tuner/documentation/tuners/). Here you can read about the input parameters for the `BayesianOptimization` tuner.

Pay special attention to these `BayesianOptimization` parameters: `num_initial_points` and `beta`. 

`num_initial_points`: 

Number of randomly selected hyperparameter combinations to try before applying bayesian probability to determine liklihood of which param combo to try next based on expected improvement


`beta`: 

Larger values means more willing to explore new hyperparameter combinations (analogous to searching for the global minimum in Gradient Descent), smaller values means that it is less willing to try new hyperparameter combinations (analogous to getting stuck in a local minimum in Gradient Descent). 

As a start, error on the side of larger values. What defines a small or large value you ask? That question would pull us into the mathematical intricacies of Bayesian Optimization and Gaussian Processes. For simplicity, notice that the default value is 2.6 and work from there. 

In [57]:
# we know that 24 samples is about 25% of 96 possible hyper-parameter combos
# because BO isn't random (after num_initial_points number of trails) let's see if 15 max trials gives good results
# feel free to play with any of these numbers
max_trials=15
num_initial_points=5
beta=5.0

In [58]:
bayesian_tuner = BayesianOptimization(
                    build_model,
                    objective='val_accuracy',
                    max_trials=max_trials,
                    hyperparameters=hp, # pass in our hyperparameter dictionary
                    num_initial_points=num_initial_points, 
                    beta=beta, 
                    seed=1234,
                    directory='./keras-tuner-trial',
                    project_name='bayesian_optimization_4')

In [59]:
bayesian_tuner.search(X_train, y_train,
               epochs=3,
               validation_data=(X_test, y_test))

Trial 15 Complete [00h 00m 07s]
val_accuracy: 0.8206400275230408

Best val_accuracy So Far: 0.8754000067710876
Total elapsed time: 00h 02m 45s
INFO:tensorflow:Oracle triggered exit


In [60]:
bayesian_tuner.results_summary()

Results summary
Results in ./keras-tuner-trial/bayesian_optimization_4
Showing 10 best trials
Objective(name='val_accuracy', direction='max')
Trial summary
Hyperparameters:
units: 352
learning_rate: 0.001
activation: relu
Score: 0.8754000067710876
Trial summary
Hyperparameters:
units: 512
learning_rate: 0.001
activation: relu
Score: 0.8732399940490723
Trial summary
Hyperparameters:
units: 480
learning_rate: 0.001
activation: sigmoid
Score: 0.8648800253868103
Trial summary
Hyperparameters:
units: 256
learning_rate: 0.001
activation: sigmoid
Score: 0.8584799766540527
Trial summary
Hyperparameters:
units: 320
learning_rate: 0.01
activation: sigmoid
Score: 0.8392400145530701
Trial summary
Hyperparameters:
units: 32
learning_rate: 0.001
activation: relu
Score: 0.8268399834632874
Trial summary
Hyperparameters:
units: 32
learning_rate: 0.001
activation: relu
Score: 0.8250799775123596
Trial summary
Hyperparameters:
units: 384
learning_rate: 0.01
activation: relu
Score: 0.8224800229072571
Trial

 ### Results
 
Identify and write the the best performing hyperparamter combination and model score. 
Note that because this is  Bayesian Optimization, multiple runs might have slighly different outcomes. 
 
 

YOUR ANSWER HERE

Best performing hyperparamter combo and score:

units: 352
learning_rate: 0.001
activation: relu
Score: 0.8754000067710876

---------
## Brute Force Gridsearch Optimization


### Populate a Sklearn compatiable parameter dictionary

In [61]:
# build out our hyperparameter dictionary 
hyper_parameters = {
    # BUG Fix: cast array as list otherwise GridSearchCV will throw error
    "units": np.arange(32, 544, 32).tolist(),
    "learning_rate": [1e-1, 1e-2, 1e-3],
    "activation":["relu", "sigmoid"]
}

In [62]:
hyper_parameters

{'units': [32,
  64,
  96,
  128,
  160,
  192,
  224,
  256,
  288,
  320,
  352,
  384,
  416,
  448,
  480,
  512],
 'learning_rate': [0.1, 0.01, 0.001],
 'activation': ['relu', 'sigmoid']}

### Build a Sklearn compatiable model function

In [63]:
def build_model(units, learning_rate, activation):
    
    """
    Returns a complie keras model ready for keras-tuner gridsearch algorithms 
    """
    
    model = Sequential()
    
    # hidden layer
    model.add(Dense(units, activation=activation))
    
    # output layer
    model.add(Dense(10, activation='softmax'))
    
    model.compile(
        optimizer=Adam(learning_rate),
        loss='sparse_categorical_crossentropy',
        metrics=['accuracy'])
    
    return model

In [64]:
model = KerasClassifier(build_fn = build_model)

In [65]:
# save start time 
start = time()

# Create Grid Search
grid = GridSearchCV(estimator=model, 
                    param_grid=hyper_parameters, 
                    n_jobs=-2, 
                    verbose=1, 
                    cv=3)

grid_result = grid.fit(X_train, y_train)

# save end time 
end = time()

# Report Results
print(f"Best: {grid_result.best_score_} using {grid_result.best_params_}")

means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']

for mean, stdev, param in zip(means, stds, params):
    print(f"Means: {mean}, Stdev: {stdev} with: {param}")

Fitting 3 folds for each of 96 candidates, totalling 288 fits


[Parallel(n_jobs=-2)]: Using backend LokyBackend with 7 concurrent workers.
2021-07-28 15:00:47.122910: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-07-28 15:00:47.122895: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-07-28 15:00:47.500294: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations: 

1118/1563 [====================>.........] - ETA: 1s - loss: 2.1067 - accuracy: 0.3655

[Parallel(n_jobs=-2)]: Done  36 tasks      | elapsed:  1.7min


782/782 [==============================] - 3s 3ms/step - loss: 0.6861 - accuracy: 0.7948


782/782 [==============================] - 2s 2ms/step - loss: 0.5482 - accuracy: 0.8365


782/782 [==============================] - 3s 3ms/step - loss: 1.2672 - accuracy: 0.6587


1373/1563 [=========================>....] - ETA: 0s - loss: 1.4217 - accuracy: 0.6095

[Parallel(n_jobs=-2)]: Done 186 tasks      | elapsed:  6.6min


782/782 [==============================] - 5s 5ms/step - loss: 0.5908 - accuracy: 0.8244


782/782 [==============================] - 4s 4ms/step - loss: 0.6890 - accuracy: 0.7962


782/782 [==============================] - 2s 2ms/step - loss: 0.6697 - accuracy: 0.8028


[Parallel(n_jobs=-2)]: Done 288 out of 288 | elapsed: 12.3min finished


2344/2344 [==============================] - 8s 3ms/step - loss: 0.6117 - accuracy: 0.8184
Best: 0.8432666659355164 using {'activation': 'relu', 'learning_rate': 0.001, 'units': 512}
Means: 0.2511066645383835, Stdev: 0.05394120367958035 with: {'activation': 'relu', 'learning_rate': 0.1, 'units': 32}
Means: 0.2558933347463608, Stdev: 0.008041387373901181 with: {'activation': 'relu', 'learning_rate': 0.1, 'units': 64}
Means: 0.27626666923364, Stdev: 0.030129509293284954 with: {'activation': 'relu', 'learning_rate': 0.1, 'units': 96}
Means: 0.25540000200271606, Stdev: 0.019054809460001487 with: {'activation': 'relu', 'learning_rate': 0.1, 'units': 128}
Means: 0.29184000194072723, Stdev: 0.04609313914258724 with: {'activation': 'relu', 'learning_rate': 0.1, 'units': 160}
Means: 0.2986533244450887, Stdev: 0.011356263602729911 with: {'activation': 'relu', 'learning_rate': 0.1, 'units': 192}
Means: 0.2974933385848999, Stdev: 0.028912123202692836 with: {'activation': 'relu', 'learning_rate': 0

In [66]:
# total run time 
total_run_time_in_miniutes = (end - start)/60
total_run_time_in_miniutes

12.458556234836578

In [67]:
grid_result.best_params_

{'activation': 'relu', 'learning_rate': 0.001, 'units': 512}

In [68]:
# because all other optimization approaches are reporting test set score
# let's calculate the test set score in this case 
best_model = grid_result.best_estimator_
test_acc = best_model.score(X_test, y_test)

782/782 [==============================] - 1s 2ms/step - loss: 0.4941 - accuracy: 0.8548


In [69]:
test_acc

0.8547999858856201

 ### Results
 
Identify and write the the best performing hyperparamter combination and model score. 
 
 

YOUR ANSWER HERE

_______

# Conclusion

The spirit of this experiment is to expose you to the idea of benchmarking and comparing the trade-offs of various gridsearch approaches. 

Even if we did find a way to pass in the original test set into GridSearchCV, we can see that both Random Search and Bayesian Optimization are arguably better alternatives to a brute force grid search when we consider the trade-offs of run time and locating the best performing model. 

----

# Stretch Goals

- Feel free to run whatever gridserach experiments on whatever models you like!

In [ ]:
# this is your open playground - be free to explore as you wish 